File: src/test/java/ar/edu/unq/epers/unidad5/PersonajeServiceTest.java
package ar.edu.unq.epers.unidad5;

import ar.edu.unq.epers.unidad5.model.Dungeon;
import ar.edu.unq.epers.unidad5.model.Item;
import ar.edu.unq.epers.unidad5.model.Personaje;
import ar.edu.unq.epers.unidad5.model.PesoPromedio;
import ar.edu.unq.epers.unidad5.service.PersonajeService;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;
import java.util.stream.IntStream;

@SpringBootTest
public class PersonajeServiceTest {

    @Autowired private PersonajeService personajeService;

    @BeforeEach
    public void setup() {
        List<Dungeon> dungeons = List.of(
                new Dungeon("Palacio del infinito"),
                new Dungeon("Calabozo del androide"),
                new Dungeon("Via circuito")
        );
        List<Item> items = List.of(
                new Item("Baculo", 25, List.of(dungeons.get(0))),
                new Item("Tunica", 30, List.of(dungeons.get(1), dungeons.get(0))),
                new Item("Sombrero", 15, List.of(dungeons.get(2))),
                new Item("Pocion", 4, List.of(dungeons.get(1), dungeons.get(2)))
        );

        List<Personaje> personajes = IntStream.range(0, 1000).mapToObj(i -> {
            Personaje personaje = new Personaje("Maguito " + i, i + 1, 80);
            recogerItems(items, personaje);
            return personaje;
        }).toList();

        personajeService.crearTodos(personajes);
    }

    private void recogerItems(List<Item> items, Personaje personaje) {
        items.forEach(personaje::recoger);
    }

    @Test
    public void recuperarPorNombre() {
        List<Personaje> personajes = personajeService.recuperarPorNombre("Maguito 4");
        Personaje personaje = personajes.get(0);

        Assertions.assertEquals(5, personaje.getVida());
        Assertions.assertEquals(80, personaje.getPesoMaximo());
        Assertions.assertEquals(74, personaje.getPesoActual());
        Assertions.assertEquals(4, personaje.getInventario().size());
    }

    @Test
    public void recuperarPorVida() {
        List<Personaje> personajes = personajeService.recuperarPorVida(10);
        Personaje personaje = personajes.get(0);

        Assertions.assertEquals("Maguito 9", personaje.getNombre());
        Assertions.assertEquals(80, personaje.getPesoMaximo());
        Assertions.assertEquals(74, personaje.getPesoActual());
        Assertions.assertEquals(4, personaje.getInventario().size());
    }

    @Test
    public void recuperarPorRangoDeVida() {
        List<Personaje> personajes = personajeService.recuperarPorRangoDeVida(20, 30);
        Assertions.assertEquals(9, personajes.size());
    }

    @Test
    public void recuperarItemsDePersonaje() {
        List<Item> items = personajeService.recuperarItemsDePersonaje("Maguito 405");
        Item tunica = items.get(0);
        Item baculo = items.get(1);
        Item sombrero = items.get(2);
        Item pocion = items.get(3);

        Assertions.assertEquals("Tunica", tunica.getNombre());
        Assertions.assertEquals(30, tunica.getPeso());
        Assertions.assertEquals("Baculo", baculo.getNombre());
        Assertions.assertEquals(25, baculo.getPeso());
        Assertions.assertEquals("Sombrero", sombrero.getNombre());
        Assertions.assertEquals(15, sombrero.getPeso());
        Assertions.assertEquals("Pocion", pocion.getNombre());
        Assertions.assertEquals(4, pocion.getPeso());
    }

    @Test
    public void obtenerPesoPromedio() {
        List<PesoPromedio> pesosPromedios = personajeService.obtenerPesoPromedio(List.of("Maguito 200", "Maguito 500"));

        Assertions.assertEquals("Maguito 500", pesosPromedios.get(0).getNombre());
        Assertions.assertEquals(18.5, pesosPromedios.get(0).getValor());
        Assertions.assertEquals("Maguito 200", pesosPromedios.get(1).getNombre());
        Assertions.assertEquals(18.5, pesosPromedios.get(1).getValor());
    }

    @AfterEach
    public void deleteAll() {
        personajeService.eliminarTodo();
    }
}

File: src/test/kotlin/ar/edu/unq/epers/unidad5/PersonajeServiceTest.kt
package ar.edu.unq.epers.unidad5

import ar.edu.unq.epers.unidad5.model.Dungeon
import ar.edu.unq.epers.unidad5.model.Item
import ar.edu.unq.epers.unidad5.model.Personaje
import ar.edu.unq.epers.unidad5.service.PersonajeService
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import java.util.function.Consumer
import java.util.stream.IntStream

@SpringBootTest
class PersonajeServiceTest {

    @Autowired
    lateinit var personajeService: PersonajeService

    @BeforeEach
    fun setup() {
        val dungeons = listOf(
            Dungeon("Palacio del infinito"),
            Dungeon("Calabozo del androide"),
            Dungeon("Via circuito")
        )
        val items = listOf(
            Item("Baculo", 25, mutableListOf(dungeons[0])),
            Item("Tunica", 30, mutableListOf(dungeons[1], dungeons[0])),
            Item("Sombrero", 15, mutableListOf(dungeons[2])),
            Item("Pocion", 4, mutableListOf(dungeons[1], dungeons[2]))
        )

        val personajes = IntStream.range(0, 1000).mapToObj { i: Int ->
            val personaje = Personaje("Maguito $i", i+1, 80)
            recogerItems(items, personaje)
            personaje
        }
        personajeService.crearTodos(personajes.toList())
    }

    private fun recogerItems(items: List<Item>, personaje: Personaje) {
        items.forEach(Consumer { i: Item ->
            personaje.recoger(i)
        })
    }

    @Test
    fun recuperarPorNombre() {
        val personajes = personajeService.recuperarPorNombre("Maguito 4")
        val personaje = personajes[0]

        Assertions.assertEquals(5, personaje.vida)
        Assertions.assertEquals(80, personaje.pesoMaximo)
        Assertions.assertEquals(74, personaje.pesoActual)
        Assertions.assertEquals(4, personaje.inventario.size)
    }

    @Test
    fun recuperarPorVida() {
        val personajes = personajeService.recuperarPorVida(10)
        val personaje = personajes[0]

        Assertions.assertEquals("Maguito 9", personaje.nombre)
        Assertions.assertEquals(80, personaje.pesoMaximo)
        Assertions.assertEquals(74, personaje.pesoActual)
        Assertions.assertEquals(4, personaje.inventario.size)
    }

    @Test
    fun recuperarPorRangoDeVida() {
        val personajes = personajeService.recuperarPorRangoDeVida(20, 30)

        Assertions.assertEquals(9, personajes.size)
    }

    @Test
    fun recuperarItemsDePersonaje() {
       val items = personajeService.recuperarItemsDePersonaje("Maguito 405")
        val tunica = items[0]
        val baculo = items[1]
        val sombrero = items[2]
        val pocion = items[3]
        Assertions.assertEquals("Tunica", tunica.nombre)
        Assertions.assertEquals(30, tunica.peso)
        Assertions.assertEquals("Baculo", baculo.nombre)
        Assertions.assertEquals(25, baculo.peso)
        Assertions.assertEquals("Sombrero", sombrero.nombre)
        Assertions.assertEquals(15, sombrero.peso)
        Assertions.assertEquals("Pocion", pocion.nombre)
        Assertions.assertEquals(4, pocion.peso)
    }

    @Test
    fun obtenerPesoPromedio() {
        val pesosPromedios = personajeService.obtenerPesoPromedio(listOf("Maguito 200", "Maguito 500"))

        Assertions.assertEquals("Maguito 500", pesosPromedios[0].nombre)
        Assertions.assertEquals(18.5, pesosPromedios[0].valor)
        Assertions.assertEquals("Maguito 200", pesosPromedios[1].nombre)
        Assertions.assertEquals(18.5, pesosPromedios[1].valor)
    }

    @AfterEach
    fun deleteAll(){
        personajeService.eliminarTodo()
    }
}